### PR TITLE
Farm card redesign and redux data handling changes

### PR DESCRIPTION
**Pull request by RichardPK**
**Notes**
- Tiny UI Kit PR fixing `textTransform` style declaration should be merged and released first https://github.com/pancakeswap/pancake-uikit/pull/139
- Once reviewed there can be some tidy up of existing components that we no longer use - namely `views/Farm/Farm`, but there could be more..

**Overview**
- Modifying the existing `FarmCard` component to encapsulate most of the changes in the new designs:

Screenshot 2021-01-10 at 16 45 28

**For a later PR:**
- APY calculator
- Modal updates

**Code changes**

Data handling/redux changes:
- `fetchFarmUserDataAsync` now gets data for all farms across 4 calls (allowances, tokenBalance, stakedBalance, farmEarnings), and no longer takes in a PID
- `src/state/farms/fetchFarmUser` has the 4 separate calls as mentioned above.
- `useFarmFromPid` has been simplified to just return redux data for a specific pid - it no longer also fetches farm data for that pid. The `Farm` view itself not dispatches the `fetchFarmUserDataAsync` to populate that data.

Farm Cards:
- A lot of the old `FarmCard` broken into separate components
- `HarvestAction` and `StakingAction` responsible for the staking/harvesting buttons & calculating relevant values
- `CardActionsContainer` is responsible for rendering the 'actions' section of the card - Staking, Harvesting and account-unlocking.
- `CardHeading` is a simple FE component rendering the top section of the card
- `DetailsSection` is a simple FE component responsible for rendering information that appears when the user clicks the 'details' button

Other:
- Made the Lottery `DetailsButton` more generic, to be used for any collapsable sections - `ExpandableSectionButton`. There are also some minor changes to `TotalPrizesCard` following this
- Created a util to get the first and second parts of an LP url - `getLiquidityUrlPathParts`

### CODE REVIEW

File: src/views/Farms/Farms.tsx [hunk 1]
```diff
@@ -52,9 +67,19 @@ const Farms: React.FC = () => {
...
+          removed={removed}
+          bnbPrice={bnbPrice}
+          cakePrice={cakePrice}
+          ethereum={ethereum}
+          account={account}
+        />
+      ))
     },
-    [bnbPrice, farmsLP],
+    [bnbPrice, farmsLP, account, cakePrice, ethereum],
```
Comment (RichardPK) [inline src/views/Farms/Farms.tsx:82]: This function is pretty damn nasty to start with, I just added dependencies to the hook as they're required by the new farm cards

File: src/views/Farms/components/FarmCard/DetailsSection.tsx [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  tokenAddresses,
+}) => {
+  const TranslateString = useI18n()
+  const liquidityUrlPathParts = getLiquidityUrlPathParts({ quoteTokenAdresses, quoteTokenSymbol, tokenAddresses })
+
+  return (
+    <Wrapper>
+      <Flex justifyContent="space-between">
+        <Text>{TranslateString(316, 'Stake')}:</Text>
+        <StyledLink external href={`https://exchange.pancakeswap.finance/#/add/${liquidityUrlPathParts}`}>
```
Comment (RabbitDoge) [inline src/views/Farms/components/FarmCard/DetailsSection.tsx:56]: You have a component doing this here https://github.com/pancakeswap/pancake-uikit/blob/master/src/components/Link/LinkExternal.tsx#L6

File: src/views/Farms/components/FarmCard/DetailsSection.tsx [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        </StyledLink>
+      </Flex>
+      {!removed && (
+        <Flex justifyContent="space-between">
+          <Text>{TranslateString(23, 'Total Liquidity')}:</Text>
+          <Text>{totalValueFormated}</Text>
+        </Flex>
+      )}
+      <Flex justifyContent="flex-start">
+        <Link external href={bscScanAddress} target="_blank" bold={false}>
```
Comment (RabbitDoge) [inline src/views/Farms/components/FarmCard/DetailsSection.tsx:67]: the `target="_blank"` is already applied by the `external` props

File: src/views/Farms/components/FarmCard/FarmCard.tsx [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+import { QuoteToken } from 'config/constants/types'
+import DetailsSection from './DetailsSection'
+import CardHeading from './CardHeading'
+import CardActionsContainer from './CardActionsContainer'
+
+export interface FarmWithStakedValue extends Farm {
+  apy?: BigNumber
+}
+
+const RainbowLight = keyframes`
```
Comment (RabbitDoge) [inline src/views/Farms/components/FarmCard/FarmCard.tsx:19]: not sure why prettier is not formatting this 🤔

**Review by RabbitDoge — CHANGES_REQUESTED** [review]

File: src/views/Farms/Farms.tsx (hunk 1 above)
Comment (RichardPK) [reply src/views/Farms/Farms.tsx:82]: 👍

File: src/utils/getLiquidityUrlPathParts.ts [hunk 5]
```diff
@@ -0,0 +1,8 @@
+const getLiquidityUrlPathParts = ({ quoteTokenAdresses, quoteTokenSymbol, tokenAddresses }) => {
```
[the remark this answers is not in the document]
Comment (RichardPK) [reply src/utils/getLiquidityUrlPathParts.ts:1]: Cool - have done. Frankly I'm not 100% why ETH is used here? But it is how the 'Add Liquidity' urls are constructed nonetheless, so I mirrored it. A hangover from the sushi fork?

File: src/views/Farms/components/FarmCard/CardActionsContainer.tsx [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    ) : (
+      <Button mt="8px" fullWidth disabled={requestedApproval} onClick={handleApprove}>
+        {TranslateString(999, 'Approve Contract')}
+      </Button>
+    )
+  }
+
+  return (
+    <Action>
+      {/* TODO: These text elements should make use of updated textTransform prop when uikit version is deployed */}
```
[the remark this answers is not in the document]
Comment (RichardPK) [reply src/views/Farms/components/FarmCard/CardActionsContainer.tsx:65]: Akh annoyingly I added a bug in the `textTransform` implementation - fixed here https://github.com/pancakeswap/pancake-uikit/pull/139 (I think it was actually included by travioli in another PR, so it's in and release. I'll remove the comment)

[the remark this answers is not in the document]
Comment (RichardPK) [reply src/views/Farms/components/FarmCard/CardActionsContainer.tsx:65]: I've brought in uikit 0.9.0 which has the fix in it 👍

**Review by hachiojidev — APPROVED** [review]

File: src/utils/getLiquidityUrlPathParts.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (RabbitDoge) [reply src/utils/getLiquidityUrlPathParts.ts:4]: Yes probably from the Sushi fork

**Review by RabbitDoge — APPROVED** [review]

**Pull request merged**